BUG: publishing segments with kafka index service failed in druid 0.9.2

```
2016-12-17T22:46:33,776 WARN [task-runner-0-priority-0] io.druid.segment.realtime.appenderator.FiniteAppenderatorDriver - Failed publishAll (try 4), retrying in 14,822ms.
java.util.concurrent.ExecutionException: java.lang.IllegalAccessError: tried to access method org.apache.hadoop.fs.FileSystem.rename(Lorg/apache/hadoop/fs/Path;Lorg/apache/hadoop/fs/Path;[Lorg/apache/hadoop/fs/Options$Rename;)V from class org.apache.hadoop.fs.HadoopFsWrapper
	at com.google.common.util.concurrent.AbstractFuture$Sync.getValue(AbstractFuture.java:299) ~[guava-16.0.1.jar:?]
	at com.google.common.util.concurrent.AbstractFuture$Sync.get(AbstractFuture.java:286) ~[guava-16.0.1.jar:?]
	at com.google.common.util.concurrent.AbstractFuture.get(AbstractFuture.java:116) ~[guava-16.0.1.jar:?]
	at io.druid.segment.realtime.appenderator.FiniteAppenderatorDriver.publishAll(FiniteAppenderatorDriver.java:417) [druid-server-0.9.2.jar:0.9.2]
	at io.druid.segment.realtime.appenderator.FiniteAppenderatorDriver.finish(FiniteAppenderatorDriver.java:256) [druid-server-0.9.2.jar:0.9.2]
	at io.druid.indexing.kafka.KafkaIndexTask.run(KafkaIndexTask.java:522) [druid-kafka-indexing-service-0.9.2.jar:0.9.2]
	at io.druid.indexing.overlord.ThreadPoolTaskRunner$ThreadPoolTaskRunnerCallable.call(ThreadPoolTaskRunner.java:436) [druid-indexing-service-0.9.2.jar:0.9.2]
	at io.druid.indexing.overlord.ThreadPoolTaskRunner$ThreadPoolTaskRunnerCallable.call(ThreadPoolTaskRunner.java:408) [druid-indexing-service-0.9.2.jar:0.9.2]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_77]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [?:1.8.0_77]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [?:1.8.0_77]
	at java.lang.Thread.run(Thread.java:745) [?:1.8.0_77]
Caused by: java.lang.IllegalAccessError: tried to access method org.apache.hadoop.fs.FileSystem.rename(Lorg/apache/hadoop/fs/Path;Lorg/apache/hadoop/fs/Path;[Lorg/apache/hadoop/fs/Options$Rename;)V from class org.apache.hadoop.fs.HadoopFsWrapper
	at org.apache.hadoop.fs.HadoopFsWrapper.rename(HadoopFsWrapper.java:51) ~[?:?]
	at io.druid.storage.hdfs.HdfsDataSegmentPusher.push(HdfsDataSegmentPusher.java:119) ~[?:?]
	at io.druid.segment.realtime.appenderator.AppenderatorImpl.mergeAndPush(AppenderatorImpl.java:571) ~[druid-server-0.9.2.jar:0.9.2]
	at io.druid.segment.realtime.appenderator.AppenderatorImpl.access$600(AppenderatorImpl.java:93) ~[druid-server-0.9.2.jar:0.9.2]
	at io.druid.segment.realtime.appenderator.AppenderatorImpl$3.apply(AppenderatorImpl.java:467) ~[druid-server-0.9.2.jar:0.9.2]
	at io.druid.segment.realtime.appenderator.AppenderatorImpl$3.apply(AppenderatorImpl.java:455) ~[druid-server-0.9.2.jar:0.9.2]
	at com.google.common.util.concurrent.Futures$1.apply(Futures.java:713) ~[guava-16.0.1.jar:?]
	at com.google.common.util.concurrent.Futures$ChainingListenableFuture.run(Futures.java:861) ~[guava-16.0.1.jar:?]
	... 3 more
```

this seems related to #3650.

--- a/extensions-core/hdfs-storage/src/main/java/org/apache/hadoop/fs/HadoopFsWrapper.java
+++ b/extensions-core/hdfs-storage/src/main/java/org/apache/hadoop/fs/HadoopFsWrapper.java
@@ -19,9 +19,12 @@
 
 package org.apache.hadoop.fs;
 
+import com.google.common.base.Throwables;
 import io.druid.java.util.common.logger.Logger;
 
 import java.io.IOException;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
 
 /**
  * This wrapper class is created to be able to access some of the the "protected" methods inside Hadoop's
@@ -50,12 +53,23 @@ public class HadoopFsWrapper
   public static boolean rename(FileSystem fs, Path from, Path to, boolean replaceExisting) throws IOException
   {
     try {
-      fs.rename(from, to, replaceExisting ? Options.Rename.OVERWRITE : Options.Rename.NONE);
+      // Note: Using reflection instead of simpler
+      // fs.rename(from, to, replaceExisting ? Options.Rename.OVERWRITE : Options.Rename.NONE);
+      // due to the issues discussed in https://github.com/druid-io/druid/pull/3787
+      Method renameMethod = fs.getClass().getMethod("rename", Path.class, Path.class, Options.Rename[].class);
+      renameMethod.invoke(fs, from, to, new Options.Rename[]{Options.Rename.NONE});
       return true;
     }
-    catch (FileAlreadyExistsException ex) {
-      log.info(ex, "Destination exists while renaming [%s] to [%s]", from, to);
-      return false;
+    catch (InvocationTargetException ex) {
+      if (ex.getTargetException() instanceof FileAlreadyExistsException) {
+        log.info(ex, "Destination exists while renaming [%s] to [%s]", from, to);
+        return false;
+      } else {
+        throw Throwables.propagate(ex);
+      }
+    }
+    catch (NoSuchMethodException | IllegalAccessException ex) {
+      throw Throwables.propagate(ex);
     }
   }
 }

--- a/extensions-core/hdfs-storage/src/main/java/org/apache/hadoop/fs/HadoopFsWrapper.java
+++ b/extensions-core/hdfs-storage/src/main/java/org/apache/hadoop/fs/HadoopFsWrapper.java
@@ -56,7 +56,7 @@ public class HadoopFsWrapper
       // Note: Using reflection instead of simpler
       // fs.rename(from, to, replaceExisting ? Options.Rename.OVERWRITE : Options.Rename.NONE);
       // due to the issues discussed in https://github.com/druid-io/druid/pull/3787
-      Method renameMethod = fs.getClass().getMethod("rename", Path.class, Path.class, Options.Rename[].class);
+      Method renameMethod = findRenameMethodRecursively(fs.getClass());
       renameMethod.invoke(fs, from, to, new Options.Rename[]{Options.Rename.NONE});
       return true;
     }
@@ -69,7 +69,33 @@ public class HadoopFsWrapper
       }
     }
     catch (NoSuchMethodException | IllegalAccessException ex) {
+
+      for (Method method : fs.getClass().getDeclaredMethods()) {
+        log.error(method.toGenericString());
+      }
       throw Throwables.propagate(ex);
+    }
+  }
+
+  /**
+   * Finds "rename" method recursively through the FileSystem class hierarchy. This is required because
+   * clazz.getMethod(..) only returns PUBLIC methods in clazz hierarchy.
+   * and clazz.getDeclaredMethod(..) only returns all methods declared in clazz but not inherited ones.
+   */
+  private static Method findRenameMethodRecursively(Class<?> clazz) throws NoSuchMethodException
+  {
+    try {
+      Method renameMethod = clazz.getDeclaredMethod("rename", Path.class, Path.class, Options.Rename[].class);
+      renameMethod.setAccessible(true);
+      return renameMethod;
+    }
+    catch (NoSuchMethodException ex) {
+      Class<?> superClazz = clazz.getSuperclass();
+      if (superClazz == null) {
+        throw ex;
+      } else {
+        return findRenameMethodRecursively(superClazz);
+      }
     }
   }
 }

